### PR TITLE
Sqlite encryption

### PR DESCRIPTION
**Pull request by amit-metri**
- added Sqlite encryption.
- Sqlite encryption is added on top of shared preference encryption as it was required to store the random passphrase.

### CODE REVIEW

File: app/build.gradle [hunk 1]
```diff
@@ -156,6 +156,13 @@ dependencies {
     // In case 'lifecycleAutotracking' is enabled
     implementation "androidx.lifecycle:lifecycle-extensions:$versions.snowplow_lifecycle"
 
+    //Crypto
+    implementation "androidx.security:security-crypto:$versions.crypto"
+
+    //SqlCipher
+    implementation "net.zetetic:android-database-sqlcipher:$versions.sqlcipher"
+    implementation "androidx.sqlite:sqlite:$versions.sqlite"
```
Comment (pinakinkansara) [inline app/build.gradle:164]: change androidx.sqlite:sqlite to androidx.sqlite:sqlite-ktx

File: app/src/main/java/ca/bc/gov/bchealth/datasource/DataStoreRepo.kt [hunk 2]
```diff
@@ -1,69 +1,83 @@
...
-
 class DataStoreRepo @Inject constructor(
-    private val context: Context
+    private val encryptedPreferences: SharedPreferences
 ) {
 
     companion object {
-        val ON_BOARDING_SHOWN = booleanPreferencesKey("ON_BOARDING_SHOWN")
-        val IS_ANALYTICS_ENABLED = booleanPreferencesKey("IS_ANALYTICS_ENABLED")
+        const val ON_BOARDING_SHOWN = "ON_BOARDING_SHOWN"
```
Comment (pinakinkansara) [inline app/src/main/java/ca/bc/gov/bchealth/datasource/DataStoreRepo.kt:18]: Mark all this constants as private

File: app/src/main/java/ca/bc/gov/bchealth/di/DataStoreModule.kt [hunk 3]
```diff
@@ -20,5 +24,25 @@ class DataStoreModule {
...
-    fun providesDataStore(@ApplicationContext context: Context) = DataStoreRepo(context = context)
+    fun providesDataStore(sharedPreferences: SharedPreferences) =
+        DataStoreRepo(sharedPreferences)
+
+    @Provides
+    @Singleton
+    fun provideEncryptedPreferences(@ApplicationContext context: Context): SharedPreferences {
+
+        // Step 1: Create or retrieve the Master Key for encryption/decryption
+        val masterKey = MasterKey.Builder(context)
```
Comment (pinakinkansara) [inline app/src/main/java/ca/bc/gov/bchealth/di/DataStoreModule.kt:35]: This masterKey should be a separate provide method in the class

File: app/src/main/java/ca/bc/gov/bchealth/di/HealthCadDataSourceModule.kt [hunk 4]
```diff
@@ -24,13 +26,37 @@ class HealthCadDataSourceModule {
...
+            BcVaccineCardDataBase::class.java,
+            "bc_vaccine_card_db"
+        )
+            .openHelperFactory(supportFactory)
+            .addMigrations(MIGRATION_1_2, MIGRATION_2_3)
+            .fallbackToDestructiveMigration()
+            .build()
+    }
+
+    private fun getRandom(): ByteArray {
```
Comment (pinakinkansara) [inline app/src/main/java/ca/bc/gov/bchealth/di/HealthCadDataSourceModule.kt:53]: extract this logic to a separate class.

File: scripts/versions.gradle [hunk 5]
```diff
@@ -45,6 +45,11 @@ versions.logging_interceptor = "4.9.2"
 versions.snowplow = "2.2.1"
 versions.snowplow_lifecycle = "2.2.0"
 
+versions.crypto = "1.1.0-alpha03"
+
+versions.sqlcipher = "4.5.0"
+versions.sqlire = "2.1.0"
```
Comment (pinakinkansara) [inline scripts/versions.gradle:51]: variable name should be sqlite instead of sqlire.

**Review by pinakinkansara — CHANGES_REQUESTED** [review]

File: app/src/main/java/ca/bc/gov/bchealth/di/HealthCadDataSourceModule.kt [hunk 6]
```diff
@@ -24,13 +26,37 @@ class HealthCadDataSourceModule {
...
+        val allowedChars = ('A'..'Z') + ('a'..'z') + ('0'..'9')
+        return (1..32)
+            .map { allowedChars.random() }
```
Comment (will-freshworks) [inline app/src/main/java/ca/bc/gov/bchealth/di/HealthCadDataSourceModule.kt:56]: Don't use `.random` for secure password generation. You should use SecureRandom. Here's a (Java) code sample for generating a password using SecureRandom:
```
val random = new SecureRandom();
val generatedBytes = random.generateSeed(24);
val generatedPass = Base64.encodeToString(generated, Base64.DEFAULT);
```

File: app/src/main/java/ca/bc/gov/bchealth/datasource/DataStoreRepo.kt [hunk 7]
```diff
@@ -1,69 +1,83 @@
...
+    fun setRecentFormData(formData: String) {
+        encryptedPreferences.edit().putString(
+            RECENT_FORM_DATA, formData
+        ).apply()
+    }
+
+    val getPassPhrase: ByteArray = encryptedPreferences
+        .getString(PASS_PHRASE, "")?.toByteArray() ?: "".toByteArray()
+
+    fun setPassPhrase(passPhrase: ByteArray) {
```
Comment (will-freshworks) [inline app/src/main/java/ca/bc/gov/bchealth/datasource/DataStoreRepo.kt:78]: I thought getters/setters weren't recommended in Kotlin. Could you use properties?

File: app/src/main/java/ca/bc/gov/bchealth/di/HealthCadDataSourceModule.kt (hunk 6 above)
Comment (will-freshworks) [reply app/src/main/java/ca/bc/gov/bchealth/di/HealthCadDataSourceModule.kt:56]: Oh, you need a byte array, so you can just use the first two lines of this, no need to put it into a string.

**Review by will-freshworks — CHANGES_REQUESTED** [review]
Minor changes suggested, most important one is switching to securerandom

File: app/build.gradle (hunk 1 above)
Comment (amit-metri) [reply app/build.gradle:164]: Completed

File: app/src/main/java/ca/bc/gov/bchealth/datasource/DataStoreRepo.kt (hunk 2 above)
Comment (amit-metri) [reply app/src/main/java/ca/bc/gov/bchealth/datasource/DataStoreRepo.kt:18]: Completed

File: app/src/main/java/ca/bc/gov/bchealth/di/DataStoreModule.kt (hunk 3 above)
Comment (amit-metri) [reply app/src/main/java/ca/bc/gov/bchealth/di/DataStoreModule.kt:35]: Completed

File: app/src/main/java/ca/bc/gov/bchealth/di/HealthCadDataSourceModule.kt (hunk 4 above)
Comment (amit-metri) [reply app/src/main/java/ca/bc/gov/bchealth/di/HealthCadDataSourceModule.kt:53]: Completed

File: scripts/versions.gradle (hunk 5 above)
Comment (amit-metri) [reply scripts/versions.gradle:51]: Completed

File: app/src/main/java/ca/bc/gov/bchealth/di/HealthCadDataSourceModule.kt (hunk 6 above)
Comment (amit-metri) [reply app/src/main/java/ca/bc/gov/bchealth/di/HealthCadDataSourceModule.kt:56]: Completed

File: app/src/main/java/ca/bc/gov/bchealth/datasource/DataStoreRepo.kt (hunk 7 above)
Comment (amit-metri) [reply app/src/main/java/ca/bc/gov/bchealth/datasource/DataStoreRepo.kt:78]: Yes, properties are used for the above-mentioned code. For the rest of the getters and setters methods, the getter and setter object types are different. Therefore properties cannot be used for other preferences in this class.

File: app/src/main/java/ca/bc/gov/bchealth/di/HealthCadDataSourceModule.kt (hunk 6 above)
Comment (will-freshworks) [reply app/src/main/java/ca/bc/gov/bchealth/di/HealthCadDataSourceModule.kt:56]: Could you push the change so I can see it?

Comment (amit-metri) [reply app/src/main/java/ca/bc/gov/bchealth/di/HealthCadDataSourceModule.kt]: Just pushed code.

**Review by pinakinkansara — APPROVED** [review]

**Review by will-freshworks — APPROVED** [review]

**Pull request merged**